// axiom/dist/bin.js
#!/usr/bin/env node
import {
  AxiomReporter,
  EVAL_CONTEXT,
  flush,
  getEvalContext,
  initInstrumentation,
  loadConfig,
  putOnSpan,
  setAxiomConfig,
  setGlobalFlagOverrides,
  u,
  withEvalContext
} from "./chunk-WXAXKFRV.js";
import {
  AxiomCLIError
} from "./chunk-N7MOZLNM.js";
import "./chunk-MNPGB5RW.js";
import "./chunk-KEXKKQVW.js";

// src/bin.ts
import { Command as Command5 } from "commander";

// src/cli/commands/push.command.ts
import { Command } from "commander";

// src/transpiler.ts
import { build } from "esbuild";
import os from "os";
import fs from "fs/promises";
import path from "path";
import { basename } from "path";
async function loadPromptModule(filePath) {
  const result = await build({
    entryPoints: [filePath],
    bundle: true,
    write: false,
    platform: "node",
    format: "esm",
    target: ["node18"],
    sourcemap: false,
    external: [
      // Only Node.js built-ins should be external
      "fs",
      "fs/promises",
      "node:fs",
      "node:fs/promises",
      "readline",
      "node:readline",
      "path",
      "node:path",
      "os",
      "node:os",
      "url",
      "node:url",
      "util",
      "node:util",
      "crypto",
      "node:crypto",
      "events",
      "node:events",
      "stream",
      "node:stream",
      "buffer",
      "node:buffer",
      "process",
      "node:process"
    ]
  });
  const code = result.outputFiles[0].text;
  const tempDir = os.tmpdir();
  const tempFileName = `axiom-ai-prompt-${Date.now()}-${Math.random().toString(36).substring(2)}.mjs`;
  const tempFilePath = path.join(tempDir, tempFileName);
  try {
    await fs.writeFile(tempFilePath, code, "utf-8");
    const moduleUrl = `file://${tempFilePath}`;
    const module = await import(moduleUrl);
    return module.default || module;
  } finally {
    try {
      await fs.unlink(tempFilePath);
    } catch (error) {
      console.warn(`Failed to clean up temporary file ${tempFilePath}:`, error);
    }
  }
}
function convertTypeBoxArgumentsToJsonSchema(arguments_) {
  if (!arguments_ || typeof arguments_ !== "object") {
    return {
      type: "object",
      properties: {},
      required: [],
      additionalProperties: false
    };
  }
  const properties = {};
  const required = [];
  for (const [key, value] of Object.entries(arguments_)) {
    if (value && typeof value === "object" && value.type) {
      properties[key] = {
        type: value.type,
        ...value.description && { description: value.description },
        ...value.enum && { enum: value.enum },
        ...value.items && { items: value.items },
        ...value.properties && { properties: value.properties },
        ...value.required && { required: value.required }
      };
      required.push(key);
    }
  }
  return {
    type: "object",
    properties,
    required,
    additionalProperties: false
  };
}
function extractPromptFromModule(moduleContent, filePath) {
  const fileBaseName = basename(filePath, ".ts");
  const defaultId = fileBaseName.toLowerCase().replace(/[^a-z0-9]/g, "-");
  const convertedArguments = convertTypeBoxArgumentsToJsonSchema(moduleContent.arguments);
  const prompt = {
    name: moduleContent.name || "Untitled Prompt",
    slug: moduleContent.slug || defaultId,
    messages: moduleContent.messages || [],
    model: moduleContent.model,
    options: moduleContent.options,
    arguments: convertedArguments,
    id: moduleContent.id || defaultId,
    version: moduleContent.version || "1.0.0",
    // Optional fields from API response
    ...moduleContent.promptId && { promptId: moduleContent.promptId },
    ...moduleContent.description && { description: moduleContent.description }
  };
  if (!prompt.name) {
    throw new Error("Prompt must have a name");
  }
  if (!prompt.slug) {
    throw new Error("Prompt must have a slug");
  }
  if (!Array.isArray(prompt.messages)) {
    throw new Error("Prompt messages must be an array");
  }
  if (!prompt.model) {
    throw new Error("Prompt must have a model");
  }
  return prompt;
}
function transformJsonSchemaToTypeBox(schema) {
  if (schema.type === "string") {
    if (schema.enum && Array.isArray(schema.enum)) {
      const literals = schema.enum.map((value) => `Type.Literal('${value}')`).join(", ");
      const options = schema.description ? `, { description: '${schema.description}' }` : "";
      return `Type.Union([${literals}]${options})`;
    } else {
      const options = schema.description ? `{ description: '${schema.description}' }` : "";
      return `Type.String(${options})`;
    }
  }
  if (schema.type === "number" || schema.type === "integer") {
    const typeMethod = schema.type === "integer" ? "Integer" : "Number";
    const options = schema.description ? `{ description: '${schema.description}' }` : "";
    return `Type.${typeMethod}(${options})`;
  }
  if (schema.type === "boolean") {
    const options = schema.description ? `{ description: '${schema.description}' }` : "";
    return `Type.Boolean(${options})`;
  }
  if (schema.type === "array") {
    const itemsType = schema.items ? transformJsonSchemaToTypeBox(schema.items) : "Type.String()";
    const options = schema.description ? `, { description: '${schema.description}' }` : "";
    return `Type.Array(${itemsType}${options})`;
  }
  if (schema.type === "object") {
    if (schema.properties) {
      const props = Object.entries(schema.properties).map(([key, value]) => {
        const isRequired = schema.required && schema.required.includes(key);
        const propType = transformJsonSchemaToTypeBox(value);
        return `    ${key}: ${isRequired ? propType : `Type.Optional(${propType})`}`;
      }).join(",\n");
      const options = schema.description ? `, { description: '${schema.description}' }` : "";
      return `Type.Object({
${props}
  }${options})`;
    } else {
      const options = schema.description ? `{ description: '${schema.description}' }` : "";
      return `Type.Object({}${options ? `, ${options}` : ""})`;
    }
  }
  return "Type.String()";
}
function generatePromptFileFromApiResponse(apiResponse) {
  const { prompt, version } = apiResponse;
  const { data, options } = version;
  let argumentsCode = "{}";
  if (data.arguments && data.arguments.properties) {
    const argEntries = Object.entries(data.arguments.properties).map(([key, schema]) => {
      const isRequired = data.arguments.required && data.arguments.required.includes(key);
      const typeCode = transformJsonSchemaToTypeBox(schema);
      return `    ${key}: ${isRequired ? typeCode : `Type.Optional(${typeCode})`}`;
    }).join(",\n");
    if (argEntries) {
      argumentsCode = `{
${argEntries}
  }`;
    }
  }
  return `import { Type } from 'axiom/ai';

export default {
  name: '${prompt.name}',
  slug: '${prompt.slug}',
  description: '${prompt.description || ""}',
  messages: [${data.messages.map(
    (msg) => `
    {
      role: '${msg.role}',
      content: '${msg.content.replace(/'/g, "\\'")}',
    }`
  ).join(",")}
  ],
  model: '${data.model || "gpt-4"}',
  options: {
${options ? Object.entries(options).map(([key, value]) => `    ${key}: ${value}`).join(",\n") : ""}
  },
  arguments: ${argumentsCode},
  version: '${version.version}',
  promptId: '${prompt.promptId}',
};
`;
}

// src/cli/commands/push.command.ts
import fs2 from "fs/promises";
import readline from "readline";
async function askConfirmation(message) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  return new Promise((resolve3) => {
    rl.question(`${message} (y/N): `, (answer) => {
      rl.close();
      resolve3(answer.toLowerCase() === "y" || answer.toLowerCase() === "yes");
    });
  });
}
var loadPushCommand = (program2) => {
  const push = new Command("push").description("Push a new version of an object").argument(
    "<object>",
    "The object to push, could be a prompt, en eval, a monitor, a dashboard, etc."
  ).option("--prod", "Adds the production tag to the prompt").option("--yes", "Automatically confirm overwriting the file with server response").action(async (filePath, options) => {
    let content = null;
    if (!filePath.endsWith(".prompt.ts")) {
      console.error("Prompt files must end with .prompt.ts");
      process.exit(1);
    }
    try {
      const moduleContent = await loadPromptModule(filePath);
      const promptData = extractPromptFromModule(moduleContent, filePath);
      content = promptData;
      console.log(`Transpiled prompt: ${promptData.name} (${promptData.slug})`);
    } catch (error) {
      console.error("Failed to transpile prompt file:", error);
      process.exit(1);
    }
    if (!content) {
      console.error("No content found");
      process.exit(1);
    }
    let shouldProceed = options.yes;
    if (!shouldProceed) {
      shouldProceed = await askConfirmation(
        `This will push "${content.name}" to Axiom and overwrite ${filePath}, are you sure you want to continue?`
      );
    }
    if (!shouldProceed) {
      console.log("Push operation cancelled.");
      process.exit(0);
    }
    try {
      const response = await fetch(`${process.env.AXIOM_URL}/v1/prompts`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${process.env.AXIOM_TOKEN}`,
          "Content-Type": "application/json",
          "x-axiom-client": "axiom-ai-cli",
          "x-axiom-check": "good"
        },
        body: JSON.stringify({
          ...content,
          tags: options.yes ? ["production"] : []
        })
      });
      if (!response.ok) {
        try {
          const errorText = await response.clone().json();
          console.error(`Failed to fetch prompt: ${response.status} ${response.statusText}`);
          console.error(JSON.stringify(errorText, null, 2));
          process.exit(1);
        } catch (_error) {
          const errorText = await response.clone().text();
          console.error(`Failed to fetch prompt: ${response.status} ${response.statusText}`);
          console.error(errorText);
          process.exit(1);
        }
      }
      const apiResponse = await response.json();
      console.log(
        `Successfully pushed prompt: ${apiResponse.prompt.name} (${apiResponse.prompt.slug})`
      );
      console.log(`Version: ${apiResponse.version.version}`);
      const updatedTsContent = generatePromptFileFromApiResponse(apiResponse);
      await fs2.writeFile(filePath, updatedTsContent, "utf-8");
      console.log(`Successfully updated ${filePath}`);
    } catch (error) {
      console.error("Failed to push prompt:", error);
      process.exit(1);
    }
  });
  program2.addCommand(push);
};

// src/cli/commands/pull.command.ts
import { Command as Command2 } from "commander";
import * as fs3 from "fs/promises";
import * as path2 from "path";
var loadPullCommand = (program2) => {
  const pull = new Command2("pull").description("Pull a version of an object").argument(
    "<slug>",
    "The object to pull, could be a prompt, en eval, a monitor, a dashboard, etc."
  ).option("--version <version>", "The version to pull, default: latest", "latest").option("--output <path>", "Output file path (optional, defaults to <slug>.prompt.ts)").action(async (slug, options) => {
    try {
      console.log(`Pulling prompt: ${slug} (version: ${options.version})`);
      const url = `${process.env.AXIOM_URL}/v1/prompts/${slug}`;
      const response = await fetch(url, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${process.env.AXIOM_TOKEN}`,
          "Content-Type": "application/json",
          "x-axiom-client": "axiom-ai-cli",
          "x-axiom-check": "good"
        }
      });
      if (!response.ok) {
        try {
          const errorText = await response.clone().json();
          console.error(`Failed to fetch prompt: ${response.status} ${response.statusText}`);
          console.error(JSON.stringify(errorText, null, 2));
          process.exit(1);
        } catch (_error) {
          const errorText = await response.clone().text();
          console.error(`Failed to fetch prompt: ${response.status} ${response.statusText}`);
          console.error(errorText);
          process.exit(1);
        }
      }
      const apiResponse = await response.json();
      const tsContent = generatePromptFileFromApiResponse(apiResponse);
      const outputPath = options.output || `${slug}.prompt.ts`;
      const fullPath = path2.resolve(outputPath);
      await fs3.writeFile(fullPath, tsContent, "utf-8");
      console.log(`Successfully generated prompt file: ${fullPath}`);
      console.log(`Prompt: ${apiResponse.prompt.name} (${apiResponse.prompt.slug})`);
      console.log(`Version: ${apiResponse.version.version}`);
    } catch (error) {
      console.error("Failed to pull prompt:", error);
      process.exit(1);
    }
  });
  program2.addCommand(pull);
};

// src/cli/commands/eval.command.ts
import { Command as Command3, Argument } from "commander";

// src/evals/run-vitest.ts
import { createVitest, registerConsoleShortcuts } from "vitest/node";
var runVitest = async (dir, opts) => {
  setAxiomConfig(opts.config);
  await initInstrumentation({
    enabled: !opts.debug,
    config: opts.config
  });
  const providedConfig = {
    ...opts.config,
    eval: {
      ...opts.config.eval,
      // function can't be serialized, so we need to remove it
      instrumentation: null
    }
  };
  if (opts.debug) {
    console.log(u.bgWhite(u.blackBright(" Debug mode enabled ")));
  }
  const vi = await createVitest("test", {
    root: dir ? dir : process.cwd(),
    mode: "test",
    include: opts.include,
    exclude: opts.exclude,
    testNamePattern: opts.testNamePattern,
    reporters: ["verbose", new AxiomReporter()],
    environment: "node",
    browser: void 0,
    watch: opts.watch,
    name: "axiom:eval",
    printConsoleTrace: true,
    silent: false,
    disableConsoleIntercept: true,
    testTimeout: opts.config?.eval?.timeoutMs || 6e4,
    globals: true,
    provide: {
      baseline: opts.baseline,
      debug: opts.debug,
      overrides: opts.overrides,
      axiomConfig: providedConfig
    }
  });
  await vi.start();
  const dispose = registerConsoleShortcuts(vi, process.stdin, process.stdout);
  if (!vi.shouldKeepServer()) {
    dispose();
    await flush();
    await vi.close();
    process.exit(0);
  }
  await flush();
};

// src/cli/commands/eval.command.ts
import { lstatSync } from "fs";

// src/context.ts
function overrideFlags(partial) {
  const current = getEvalContext();
  if (!current) {
    if (process.env.NODE_ENV !== "test") {
      console.warn("overrideFlags called outside of evaluation context");
    }
    return;
  }
  const overlayContext = {
    ...current,
    flags: { ...current.flags, ...partial },
    // Merge for backwards compatibility
    parent: current,
    overrides: { ...partial }
  };
  const currentCtx = EVAL_CONTEXT.get();
  if (currentCtx) {
    Object.assign(currentCtx, overlayContext);
  }
  for (const [key, value] of Object.entries(partial)) {
    putOnSpan("flag", key, value);
  }
}

// src/cli/utils/eval-context-runner.ts
async function runEvalWithContext(overrides2, runFn) {
  setGlobalFlagOverrides(overrides2);
  return withEvalContext({ initialFlags: overrides2 }, async () => {
    if (Object.keys(overrides2).length > 0) {
      overrideFlags(overrides2);
    }
    return runFn();
  });
}

// src/cli/utils/glob-utils.ts
function isGlob(str) {
  return /[*?[\]{}!]/.test(str);
}

// src/cli/commands/eval.command.ts
var loadEvalCommand = (program2, flagOverrides = {}) => {
  return program2.addCommand(
    new Command3("eval").description("run evals locally").addArgument(
      new Argument("[target]", "file, directory, glob pattern, or eval name").default(
        ".",
        "any *.eval.ts file in current directory"
      )
    ).option("-w, --watch true", "keep server running and watch for changes", false).option("-t, --token <TOKEN>", "axiom token", process.env.AXIOM_TOKEN).option("-d, --dataset <DATASET>", "axiom dataset name", process.env.AXIOM_DATASET).option("-u, --url <AXIOM URL>", "axiom url", process.env.AXIOM_URL ?? "https://api.axiom.co").option("-b, --baseline <BASELINE ID>", "id of baseline evaluation to compare against").option("--debug", "run locally without sending to Axiom or loading baselines", false).action(async (target, options) => {
      try {
        if (options.debug) {
          process.env.AXIOM_DEBUG = "true";
        }
        let include = [];
        let exclude;
        let testNamePattern;
        const isGlobPattern = isGlob(target);
        const { config } = await loadConfig(".");
        if (isGlobPattern) {
          include = [target];
        } else {
          try {
            const stat = lstatSync(target);
            if (stat.isDirectory()) {
              include = config?.eval?.include || [];
            } else {
              include = [target];
            }
          } catch {
            testNamePattern = new RegExp(target, "i");
            include = config?.eval?.include || [];
          }
        }
        exclude = config?.eval?.exclude;
        if (!config?.eval?.instrumentation) {
          console.warn(
            u.yellow(
              "\u26A0 App instrumentation (`eval.instrumentation` in `axiom.config.ts`) not configured. Using default provider."
            )
          );
          console.log("");
        }
        await runEvalWithContext(flagOverrides, async () => {
          return runVitest(".", {
            watch: options.watch,
            baseline: options.baseline,
            include,
            exclude,
            testNamePattern,
            debug: options.debug,
            overrides: flagOverrides,
            config
          });
        });
      } catch (error) {
        if (error instanceof AxiomCLIError) {
          console.error(`
\u274C ${error.message}
`);
          process.exit(1);
        }
        throw error;
      }
    })
  );
};

// src/cli/utils/parse-flag-overrides.ts
import "zod";
import { readFileSync } from "fs";
import { resolve as resolve2 } from "path";
var FLAG_RE = /^--flag\.([^=]+)(?:=(.*))?$/;
var CONFIG_RE = /^--flags-config(?:=(.*))?$/;
function ensureNoSpaceSeparatedSyntax(flagName, value, nextToken, flagType) {
  if (value === void 0 && nextToken !== void 0) {
    if (flagType === "flag" && !nextToken.startsWith("-") && nextToken !== "true" && nextToken !== "false") {
      console.error(`\u274C Invalid syntax: --flag.${flagName} ${nextToken}`);
      console.error(`\u{1F4A1} Use: --flag.${flagName}=${nextToken}`);
      process.exit(1);
    } else if (flagType === "config" && !nextToken.startsWith("-")) {
      console.error(`\u274C Invalid syntax: --flags-config ${nextToken}`);
      console.error(`\u{1F4A1} Use: --flags-config=${nextToken}`);
      process.exit(1);
    }
  }
}
function coerceValue(raw) {
  if (raw === "true") return true;
  if (raw === "false") return false;
  const num = Number(raw);
  if (!Number.isNaN(num) && raw.trim() === num.toString()) {
    return num;
  }
  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}
function loadConfigFile(path3) {
  const abs = resolve2(process.cwd(), path3);
  try {
    const contents = readFileSync(abs, "utf8");
    const parsed = JSON.parse(contents);
    if (typeof parsed !== "object" || Array.isArray(parsed) || parsed === null) {
      console.error(
        `\u274C Flags config must be a JSON object, got ${Array.isArray(parsed) ? "array" : typeof parsed}`
      );
      process.exit(1);
    }
    return parsed;
  } catch (err) {
    console.error(`\u274C Could not read or parse flags config "${path3}": ${err.message}`);
    process.exit(1);
  }
}
function extractOverrides(argv) {
  const cleanedArgv2 = [];
  const overrides2 = {};
  let configPath = null;
  let hasCliFlags = false;
  let configPathCount = 0;
  for (let i = 0; i < argv.length; i++) {
    const token = argv[i];
    const configMatch = token.match(CONFIG_RE);
    const flagMatch = token.match(FLAG_RE);
    if (configMatch) {
      configPathCount++;
      if (configPathCount > 1) {
        console.error("\u274C Only one --flags-config can be supplied.");
        process.exit(1);
      }
      const value = configMatch[1];
      const nextToken = argv.length > i + 1 ? argv[i + 1] : void 0;
      ensureNoSpaceSeparatedSyntax("flags-config", value, nextToken, "config");
      if (!value) {
        console.error("\u274C --flags-config requires a file path");
        console.error("\u{1F4A1} Use: --flags-config=path/to/config.json");
        process.exit(1);
      }
      configPath = value;
    } else if (flagMatch) {
      hasCliFlags = true;
      const key = flagMatch[1];
      const value = flagMatch[2];
      const nextToken = argv.length > i + 1 ? argv[i + 1] : void 0;
      ensureNoSpaceSeparatedSyntax(key, value, nextToken, "flag");
      const finalValue = value === void 0 ? "true" : value;
      overrides2[key] = coerceValue(finalValue);
    } else {
      cleanedArgv2.push(token);
    }
  }
  if (configPath && hasCliFlags) {
    console.error("\u274C Cannot use both --flags-config and --flag.* arguments together.");
    console.error("Choose one approach:");
    console.error("  \u2022 Config file: --flags-config=my-flags.json");
    console.error("  \u2022 CLI flags: --flag.temperature=0.9 --flag.model=gpt-4o");
    process.exit(1);
  }
  if (configPath) {
    const configOverrides = loadConfigFile(configPath);
    return { cleanedArgv: cleanedArgv2, overrides: configOverrides };
  }
  return { cleanedArgv: cleanedArgv2, overrides: overrides2 };
}

// src/bin.ts
import pkg from "@next/env";

// src/cli/commands/version.command.ts
import { Command as Command4 } from "commander";
var loadVersionCommand = (program2) => {
  return program2.addCommand(
    new Command4("version").description("cli version").action(() => {
      console.log("0.22.2");
    })
  );
};

// src/bin.ts
var { loadEnvConfig } = pkg;
loadEnvConfig(process.cwd());
var { cleanedArgv, overrides } = extractOverrides(process.argv.slice(2));
var program = new Command5();
program.name("axiom").description("Axiom's CLI to manage your objects and run evals").version("0.22.2");
loadPushCommand(program);
loadPullCommand(program);
loadEvalCommand(program, overrides);
loadVersionCommand(program);
program.parse(["node", "axiom", ...cleanedArgv]);
export {
  program
};
